bigquery_etl/cli/query.py

```diff
@@ -768,73 +769,166 @@ def update(name, sql_dir, project_id):
-            click.echo(
-                f"Destination table {project_name}.{dataset_name}.{table_name} "
-                "does not exist in BigQuery. Run bqetl query schema deploy "
-                "<dataset>.<table> to create the destination table."
-            )
+                tmp_schema_file = NamedTemporaryFile()
+                schema = Schema.from_schema_file(query_file.parent / SCHEMA_FILE)
+                schema.to_json_file(Path(tmp_schema_file.name))
+                bigquery_schema = client.schema_from_json(tmp_schema_file.name)
+                tmp_table = bigquery.Table(tmp_identifier, schema=bigquery_schema)
+                client.create_table(tmp_table)
+
+                tmp_tables[identifier] = tmp_identifier
+
+                # get downstream dependencies that will be updated in the next iteration
+                dependencies = [
+                    p
+                    for k, refs in dependency_graph.items()
+                    for p in _queries_matching_name_pattern(k, sql_dir, project_id)
+                    if identifier in refs
+                ]
+
+                for d in dependencies:
+                    click.echo(f"Update downstream dependency schema for {d}")
+                    query_files.append(d)
+
+    if update_downstream:
+        # delete temporary tables
+        for _, table in tmp_tables.items():
+            client.delete_table(table, not_found_ok=True)
+
+
+def _update_query_schema(query_file, tmp_tables={}):
+    """
+    Update the schema of a specific query file.
+
+    Return True if the schema changed, False if it is unchanged.
+    """
+    if str(query_file) in SKIP:
+        click.echo(f"{query_file} dry runs are skipped. Cannot update schemas.")
+        return
 
-        partitioned_by = None
-        try:
-            metadata = Metadata.of_query_file(query_file_path)
+    query_file_path = Path(query_file)
 
-            if metadata.bigquery and metadata.bigquery.time_partitioning:
-                partitioned_by = metadata.bigquery.time_partitioning.field
-        except FileNotFoundError:
-            pass
+    # replace temporary table references
+    sql_content = query_file_path.read_text()
+
+    for orig_table, tmp_table in tmp_tables.items():
+        table_parts = orig_table.split(".")
+        for i in range(len(table_parts)):
+            if ".".join(table_parts[i:]) in sql_content:
+                sql_content = sql_content.replace(".".join(table_parts[i:]), tmp_table)
+                break
 
-        table_schema = Schema.for_table(
-            project_name, dataset_name, table_name, partitioned_by
+    try:
+        query_schema = Schema.from_query_file(query_file_path, sql_content)
+    except Exception:
+        click.echo(
+            click.style(
+                f"Cannot automatically update {query_file_path}. "
+                f"Please update {query_file_path / SCHEMA_FILE} manually.",
+                fg="red",
+            ),
+            err=True,
```

Comment:
For tables that reference themselves, like `clients_last_seen` or `clients_first_seen` the dry run fails here. I don't think there is a straightforward way to automatically update the schema for those.

Reply:
The only way forward I can think of would be adding support for "merging" the clients_daily schema.yaml content into the downstream tables. We'd probably need some additional metadata to indicate this is a reasonable thing to do, though, since that path would be fairly specific to the clients_last_seen problem.